### PR TITLE
Adding streaming http support.

### PR DESCRIPTION
**Pull request by handsomezebra**
This pack adds the streaming support for http.
The responses lib doesn't support mock of streaming http so I didn't have a good way to unit test it. Any suggestions?

### CODE REVIEW

**Comment by menshikh-iv** [comment]
Hi @handsomezebra, I think you can mock it manually with standart python [mock](https://docs.python.org/3/library/unittest.mock.html).

CC: @mpenkov

File: smart_open/http.py [hunk 1]
```diff
@@ -75,32 +84,42 @@ def detach(self):
...
+    def read(self, size=-1):
         """
         Mimics the read call to a filehandle object.
         """
-        logger.debug('read: %r', locals())
-        if size is None:
-            return self.response.raw.read()
+        logger.debug("reading with size: %d", size)
+        if self.response is None:
+            return ''
```
Comment (menshikh-iv) [inline smart_open/http.py]: shouldn't this be an `b''` instead of string (here and below)?

File: smart_open/http.py [hunk 2]
```diff
@@ -114,3 +133,118 @@ def readinto(self, b):
             return 0
         b[:len(data)] = data
         return len(data)
+
+
+class SeekableBufferedInputBase(BufferedInputBase):
```
Comment (menshikh-iv) [inline smart_open/http.py]: Maybe better use a more specific name (in case if this applied only for HTTP)?

Comment (mpenkov) [reply smart_open/http.py]: IMHO, this is already part of the http module, so everything in there is specific to http a priori, so the name is fine. We use a similar naming scheme in the S3 module as well.

File: smart_open/http.py [hunk 3]
```diff
@@ -75,32 +84,42 @@ def detach(self):
...
-        logger.debug('read: %r', locals())
-        if size is None:
-            return self.response.raw.read()
+        logger.debug("reading with size: %d", size)
+        if self.response is None:
+            return ''
+
+        if size == 0:
+            return ''
+        elif size < 0:
```
Comment (mpenkov) [inline smart_open/http.py]: Unless there's a good reason to nest conditionals here, I think rewriting them as un-nested is better.

```python
elif size < 0 and not self._read_buffer:
    retval = self.response.raw.read()
elif size < 0:
    retval = self._read_buffer + self.response.raw.read()
    self._read_buffer = b''
```

File: smart_open/http.py [hunk 4]
```diff
@@ -114,3 +133,118 @@ def readinto(self, b):
...
+        if whence == START:
+            new_pos = offset
+        elif whence == CURRENT:
+            new_pos = self._current_pos + offset
+        elif whence == 2:
+            new_pos = self.content_length + offset
+
+        new_pos = _clamp(new_pos, 0, self.content_length)
+
+        if self._current_pos != new_pos:
```
Comment (mpenkov) [inline smart_open/http.py]: I'd use an early return here. Saves the reader from remembering they're inside a conditional, and saves the writer an extra level of indentation. Also dodges the need to nest conditionals.

```python
if self._current_pos == new_pos:
    return self._current_pos
```

**Comment by menshikh-iv** [comment]
ping @handsomezebra, are you planning to finish PR?
CC @mpenkov

**Comment by mpenkov** [comment]
Still need to add tests.

**Comment by mpenkov** [comment]
@handsomezebra I cleaned up your commit for you, added some tests, and merged this in. Congrats on your first PR to smart_open :hooray:

**Pull request merged**